sql/catalog, backupccl: support offline descriptors, add all new restored descriptors in offline state

thoszhang opened the pull request:
Closes #53375.

----

backupccl: fix bug in writing database schema map

Release justification: Bug fixes to new functionality

Release note (bug fix): Fixes a bug introduced in a 20.2 alpha release
where incorrect caching could incur extra writes to the store during
RESTORE on user-defined schemas.

----

backupccl: reset Version and ModificationTime on new descriptors

The `Version` and `ModificationTime` fields are now reset to 1 and the
empty timestamp, respectively, on new descriptors created in RESTORE, to
make them behave like other newly created descriptors.

Release justification: Low risk, high benefit changes to existing
functionality

Release note: None

----

sql/catalog: add descriptor API support for offline state

This change adds methods for the offline state to the descriptor
interfaces and makes the state-related API more uniform.

Release justification: Low risk, high benefit changes to existing
functionality

Release note: None

----

sql/catalog: handle dropped and offline descriptors in lookups

This change updates the name resolution and descriptor lookup
implementations to respect the state of dropped and offline descriptors:
All lookups now read the `IncludeOffline` and `IncludeDropped` flags in
`tree.CommonLookupFlags` and filter results accordingly.

Tests in the context of RESTORE are forthcoming in the following commit.

Release justification: Bug fixes and low-risk updates to new
functionality

Release note: None

----

backupccl: add all new restored descriptors in offline state

This change updates the RESTORE job to add all new descriptors prior to
restoring data in the OFFLINE state, and update their state to PUBLIC
after data has been restored. This makes all descriptors behave like
tables.

Release justification: Low risk, high benefit changes to existing
functionality (for database descriptors; other changes apply only to new
functionality)

Release note (sql change): Databases being restored will now be in the
offline state, invisible to users, until the data has been restored.
This is the same as the existing behavior for restored tables. (This
change is also applied to enums and user-defined schemas being restored,
which is a change relative to only the 20.2 alpha releases.)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53539)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> Reviewed 1 of 1 files at r1, 6 of 6 files at r2, 15 of 15 files at r3, 6 of 6 files at r4, 2 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53539)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 6093 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkyb347Ubk5RZ1vPEo:-MGkyb347Ubk5RZ1vPEp:b-zah7te) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6d8a6a21cb8124e7666d7fd0dbd2ad6ab69a6/pkg/ccl/backupccl/backup_test.go#L6093)):*
<details><summary><i alt="summary">Quoted 7 lines of code…</i></summary>

> ```
> 	// Protects beforePublishingNotification, continueNotification.
> 	var mu syncutil.Mutex
> 	// Closed when the restore job has reached the point right before publishing
> 	// descriptors.
> 	beforePublishingNotification := make(chan struct{})
> 	// Closed when we're ready to resume with the restore.
> 	continueNotification := make(chan struct{})
> ```
</details>

total nit: I usually like to write this as an anonymous struct to make it clearer what the mutex protects

```
var state = struct {
    syncutil.Mutex
    beforePublishingNotification chan struct{}
    continueNotification chan struct{}   
}
```
___
*[pkg/ccl/backupccl/backup_test.go, line 6131 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkz8Zv8F9lh6rtOfCk:-MGkz8Zw7pYgczplxGGS:btqlcy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6d8a6a21cb8124e7666d7fd0dbd2ad6ab69a6/pkg/ccl/backupccl/backup_test.go#L6131)):*
> ```Go
> 		for _, server := range tc.Servers {
> 			registry := server.JobRegistry().(*jobs.Registry)
> 			registry.TestingResumerCreationKnobs = map[jobspb.Type]func(raw jobs.Resumer) jobs.Resumer{
> ```

it feels lucky that this isn't racy. 
___
*[pkg/ccl/backupccl/restore_job.go, line 1498 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGl--_X61y4Dz_B5CWS:-MGl--_X61y4Dz_B5CWT:b-pghrfm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6d8a6a21cb8124e7666d7fd0dbd2ad6ab69a6/pkg/ccl/backupccl/restore_job.go#L1498)):*
> ```
> 	// TODO (lucy): It's possible for tables and types to be added under the
> 	// schema after it becomes public. To account for this, we need to refrain
> 	// from dropping the schema if it has children, like we do for databases.
> ```
What's the scenario where we fail and want to rollback after we've made the descriptors public? 
___
*[pkg/ccl/backupccl/restore_planning.go, line 717 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkunlH-j9HoTi-X0Le:-MGkunlI4M3NsuFAoCrf:b-39puxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a5499efff628ff6ffa47fc243fa675adb8c05d9/pkg/ccl/backupccl/restore_planning.go#L717)):*
> ```Go
> 
> // rewriteDatabaseDescs rewrites all ID's in the input slice of
> // DatabaseDescriptors using the input ID rewrite mapping.
> ```

nit: comment about the semantics of and reason for `resetVersion`
___
*[pkg/ccl/backupccl/restore_planning.go, line 1507 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkvkV8A9HAsIJZc6Vs:-MGkvkV8A9HAsIJZc6Vt:b-2573yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a5499efff628ff6ffa47fc243fa675adb8c05d9/pkg/ccl/backupccl/restore_planning.go#L1507)):*
> ```Go
> 		return err
> 	}
> 	if err := rewriteDatabaseDescs(databases, descriptorRewrites, false /* resetVersion */); err != nil {
> ```

The reason this is false isn't immediately obvious to me. Could you add some commentary?
___
*[pkg/ccl/storageccl/key_rewriter.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkvQvADjFIhm5iPONh:-MGkvQvADjFIhm5iPONi:b-u9m326) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a5499efff628ff6ffa47fc243fa675adb8c05d9/pkg/ccl/storageccl/key_rewriter.go#L70)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 		// Use GetTable() (which does not check and set the ModificationTime on the
> 		// descriptor) because these table descriptors are only used for their index
> 		// information in the KeyRewriter.
> 		table := desc.GetTable()
> ```
</details>

Is this really necessary? Isn't it the case that if the version is <= 1 we don't care about the modification time being set? When is the modification time unset and the version not 0 or 1? 
___
*[pkg/sql/catalog/catalogkv/physical_accessor.go, line 81 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkwwqQ5ffbQjPhgpIt:-MGkwwqQ5ffbQjPhgpIu:b7owohe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b2dfb957aca1aba9b6f313b3ef03307b6405aad/pkg/sql/catalog/catalogkv/physical_accessor.go#L81)):*
> ```
> 	if inUnacceptableState :=
> 		(db.Dropped() && !flags.IncludeDropped) ||
> 			(db.Offline() && !flags.IncludeOffline); inUnacceptableState {
> ```
nit: I could see pulling this into a helper`func descriptorInAcceptableState(desc catalog.Descriptor, flags tree.CommonLookupFlags) bool` so this and elsewhere could be:

`if !descriptorInAcceptableState(db, flag.CommonLookupFlags) { ... }`
___
*[pkg/sql/catalog/catalogkv/physical_accessor.go, line 320 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkxbXB4CcomuH_dP7_:-MGkxbXB4CcomuH_dP7a:b-n144lw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b2dfb957aca1aba9b6f313b3ef03307b6405aad/pkg/sql/catalog/catalogkv/physical_accessor.go#L320)):*
> ```Go
> 		// return a descriptor during draining.
> 		// TODO (lucy): Is this check actually a good idea? If so, we should extend
> 		// it to all the other descriptors.
> ```

So yeah, what's the deal here? 
___
*[pkg/sql/catalog/descpb/descriptor.go, line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkwLfG5D4QQ5uEOzlv:-MGkwLfHDuk_PZdAd3uX:b-dqqg3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/794f272e6c22d5c93e63c8c09e8e0de944a2e15a/pkg/sql/catalog/descpb/descriptor.go#L98)):*
> ```
> // TODO (lucy): Does this method belong on Descriptor? This state does matter
> // for descriptor leasing, but arguably we should be upwrapping the descriptor
> // to get it.
> ```
Maybe nix this TODO and note when it makes sense to use this? 


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> Reviewed 1 of 1 files at r1, 6 of 6 files at r2, 15 of 15 files at r3, 6 of 6 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53539)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/ccl/backupccl/restore_job.go, line 1498 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGl--_X61y4Dz_B5CWS:-MGomCHqDeyfghNreN4W:b-aawwoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6d8a6a21cb8124e7666d7fd0dbd2ad6ab69a6/pkg/ccl/backupccl/restore_job.go#L1498)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// TODO (lucy): It's possible for tables and types to be added under the
> 	// schema after it becomes public. To account for this, we need to refrain
> 	// from dropping the schema if it has children, like we do for databases.
> ```
What's the scenario where we fail and want to rollback after we've made the descriptors public? 
</blockquote></details>

Cluster restore will mark all tables as public after restoring all of the user data, but before restoring the system data. This is unfortunate since if the job were to fail restoring the system data, we'd need to clean up these public tables. We should really only be marking the temporary system tables as public before the end of the job.

More generally, the Resumer could stop at any point during the job, and could enter the OnFailOrCancel hook. That being said, since the job needs to detect that it has been canceled I can only imagine this happening if:
1. the coordinator crashes after marking the tables public, but before succeeding the job
2. the job then gets canceled while it's trying to be adopted


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53539)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/ccl/backupccl/restore_job.go, line 1498 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGl--_X61y4Dz_B5CWS:-MGsDR8-AOrVgGZA7dNw:b-qbviiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6d8a6a21cb8124e7666d7fd0dbd2ad6ab69a6/pkg/ccl/backupccl/restore_job.go#L1498)):*
> That being said, since the job needs to detect that it has been canceled I can only imagine this happening if:

This isn't how it works anymore. Since 20.1, the job's context will get canceled


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53539)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 6093 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkyb347Ubk5RZ1vPEo:-MGsQryp0y19KdvLiPy_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6d8a6a21cb8124e7666d7fd0dbd2ad6ab69a6/pkg/ccl/backupccl/backup_test.go#L6093)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// Protects beforePublishingNotification, continueNotification.
> 	var mu syncutil.Mutex
> 	// Closed when the restore job has reached the point right before publishing
> 	// descriptors.
> 	beforePublishingNotification := make(chan struct{})
> 	// Closed when we're ready to resume with the restore.
> 	continueNotification := make(chan struct{})
> ```
total nit: I usually like to write this as an anonymous struct to make it clearer what the mutex protects

```
var state = struct {
    syncutil.Mutex
    beforePublishingNotification chan struct{}
    continueNotification chan struct{}   
}
```
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_test.go, line 6131 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkz8Zv8F9lh6rtOfCk:-MGsS9Z5DLeYEyEBFNaN:b119opy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6d8a6a21cb8124e7666d7fd0dbd2ad6ab69a6/pkg/ccl/backupccl/backup_test.go#L6131)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

it feels lucky that this isn't racy. 
</blockquote></details>

Is this more racy than other testing knobs we have?
___
*[pkg/ccl/backupccl/restore_job.go, line 1498 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGl--_X61y4Dz_B5CWS:-MGsSZGj4ZWOnqgoGi0q:bv01ofa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6d8a6a21cb8124e7666d7fd0dbd2ad6ab69a6/pkg/ccl/backupccl/restore_job.go#L1498)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> That being said, since the job needs to detect that it has been canceled I can only imagine this happening if:

This isn't how it works anymore. Since 20.1, the job's context will get canceled
</blockquote></details>

Yeah, what Paul said re: restoring the system data later. (With that said, why can't we mark the user tables and system tables as public at the same time? Not saying we should do that now.)
___
*[pkg/ccl/backupccl/restore_planning.go, line 717 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkunlH-j9HoTi-X0Le:-MGsg0035NmkzJ44z6d_:br52dni) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a5499efff628ff6ffa47fc243fa675adb8c05d9/pkg/ccl/backupccl/restore_planning.go#L717)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: comment about the semantics of and reason for `resetVersion`
</blockquote></details>

I'm removing this argument. My original reasoning was that if we were in the mixed-version state, we couldn't be restoring backups from a post-finalization state (but, actually, I'm not even sure if this is guaranteed to be true), so the version would be 0 and we'd want to keep it that way. But I think it's actually fine to unconditionally set the new version to 1 even in the mixed-version state. I can't think of any obstacles to correctly leasing the descriptor.
___
*[pkg/ccl/backupccl/restore_planning.go, line 1507 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkvkV8A9HAsIJZc6Vs:-MGsi-Ed1mKx3sVtiAFU:b-qdfzp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a5499efff628ff6ffa47fc243fa675adb8c05d9/pkg/ccl/backupccl/restore_planning.go#L1507)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The reason this is false isn't immediately obvious to me. Could you add some commentary?
</blockquote></details>

FWIW, this was an arbitrary choice since the only reason to call `rewrite...()` while planning is to detect errors in cross-references when assigning IDs.
___
*[pkg/ccl/storageccl/key_rewriter.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkvQvADjFIhm5iPONh:-MGsmGSa6qftSR7i6eur:btkhcqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a5499efff628ff6ffa47fc243fa675adb8c05d9/pkg/ccl/storageccl/key_rewriter.go#L70)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		// Use GetTable() (which does not check and set the ModificationTime on the
> 		// descriptor) because these table descriptors are only used for their index
> 		// information in the KeyRewriter.
> 		table := desc.GetTable()
> ```
Is this really necessary? Isn't it the case that if the version is <= 1 we don't care about the modification time being set? When is the modification time unset and the version not 0 or 1? 
</blockquote></details>

Yeah, you're right. I originally added this prior to resetting the descriptor versions and forgot that we allow an empty ModificationTime on version 1.
___
*[pkg/sql/catalog/catalogkv/physical_accessor.go, line 81 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkwwqQ5ffbQjPhgpIt:-MGspO8FB9EVpfiOluC4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b2dfb957aca1aba9b6f313b3ef03307b6405aad/pkg/sql/catalog/catalogkv/physical_accessor.go#L81)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	if inUnacceptableState :=
> 		(db.Dropped() && !flags.IncludeDropped) ||
> 			(db.Offline() && !flags.IncludeOffline); inUnacceptableState {
> ```
nit: I could see pulling this into a helper`func descriptorInAcceptableState(desc catalog.Descriptor, flags tree.CommonLookupFlags) bool` so this and elsewhere could be:

`if !descriptorInAcceptableState(db, flag.CommonLookupFlags) { ... }`
</blockquote></details>

Done.
___
*[pkg/sql/catalog/catalogkv/physical_accessor.go, line 320 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkxbXB4CcomuH_dP7_:-MGtQ3fV1m78MZGmprwb:bcfuer2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b2dfb957aca1aba9b6f313b3ef03307b6405aad/pkg/sql/catalog/catalogkv/physical_accessor.go#L320)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

So yeah, what's the deal here? 
</blockquote></details>

I think this is legit and we should do it for all the descriptor types. I'm going to do this in a separate PR, though, if you don't mind. I wrote a test for the table case, so I'll just add changes on top of that (https://github.com/cockroachdb/cockroach/pull/54213).
___
*[pkg/sql/catalog/descpb/descriptor.go, line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkwLfG5D4QQ5uEOzlv:-MGsmmav2QLKtAmxAjMF:b-ljn3me) ([raw file](https://github.com/cockroachdb/cockroach/blob/794f272e6c22d5c93e63c8c09e8e0de944a2e15a/pkg/sql/catalog/descpb/descriptor.go#L98)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> // TODO (lucy): Does this method belong on Descriptor? This state does matter
> // for descriptor leasing, but arguably we should be upwrapping the descriptor
> // to get it.
> ```
Maybe nix this TODO and note when it makes sense to use this? 
</blockquote></details>

Done. I put the doc comment on `GetDescriptorMetadata`.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 30 of 30 files at r6, 4 of 4 files at r7, 15 of 15 files at r8, 7 of 7 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-:-MGvnqCL6Ma7qUOHFRlJ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/ccl/backupccl/backup_test.go, line 6131 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/53539#-MGkz8Zv8F9lh6rtOfCk:-MGvn0UlDmAyXH7gfS6c:b-mtzqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6d8a6a21cb8124e7666d7fd0dbd2ad6ab69a6/pkg/ccl/backupccl/backup_test.go#L6131)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is this more racy than other testing knobs we have?
</blockquote></details>

fair enough


<!-- Sent from Reviewable.io -->

pbardea commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -900,10 +904,20 @@ func createImportingDescriptors(
+	for _, desc := range schemasToWrite {
+		desc.SetOffline("restoring")
+	}
+	if r.settings.Version.IsActive(ctx, clusterversion.VersionLeasedDatabaseDescriptors) {
```
A small comment here about how we need this check since 20.1 nodes don't update databases from OFFLINE to PUBLIC would be useful.

lucy-zhang replied:
Done.

pbardea reviewed (approved):
LGTM

Commit:
backupccl: reset Version and ModificationTime on new descriptors

The `Version` and `ModificationTime` fields are now reset to 1 and the
empty timestamp, respectively, on new descriptors created in RESTORE, to
make them behave like other newly created descriptors.

Release justification: Low risk, high benefit changes to existing
functionality

Release note: None

Commit:
backupccl: add all new restored descriptors in offline state

This change updates the RESTORE job to add all new descriptors prior to
restoring data in the OFFLINE state, and update their state to PUBLIC
after data has been restored. This makes all descriptors behave like
tables.

Release justification: Low risk, high benefit changes to existing
functionality (for database descriptors; other changes apply only to new
functionality)

Release note (sql change): Databases being restored will now be in the
offline state, invisible to users, until the data has been restored.
This is the same as the existing behavior for restored tables. (This
change is also applied to enums and user-defined schemas being restored,
which is a change relative to only the 20.2 alpha releases.)

Commit:
sql/catalog: add descriptor API support for offline state

This change adds methods for the offline state to the descriptor
interfaces and makes the state-related API more uniform.

Release justification: Low risk, high benefit changes to existing
functionality

Release note: None

Commit:
backupccl: fix bug in writing database schema map

Release justification: Bug fixes to new functionality

Release note (bug fix): Fixes a bug introduced in a 20.2 alpha release
where incorrect caching could incur extra writes to the store during
RESTORE on user-defined schemas.

Commit:
sql/catalog: handle dropped and offline descriptors in lookups

This change updates the name resolution and descriptor lookup
implementations to respect the state of dropped and offline descriptors:
All lookups now read the `IncludeOffline` and `IncludeDropped` flags in
`tree.CommonLookupFlags` and filter results accordingly.

Tests in the context of RESTORE are forthcoming in the following commit.

Release justification: Bug fixes and low-risk updates to new
functionality

Release note: None

lucy-zhang commented:
TFTRs

bors r+

lucy-zhang commented:
I'll wait until https://github.com/cockroachdb/cockroach/pull/53928 gets backported to backport this. cc @pbardea
